src/main/kotlin/com/qupaya/brevo/NewsletterRegistrationEventListenerProvider.kt

```diff
@@ -0,0 +1,93 @@
+package com.qupaya.brevo
+
+import org.jboss.logging.Logger
+import org.keycloak.events.Event
+import org.keycloak.events.EventListenerProvider
+import org.keycloak.events.EventType
+import org.keycloak.events.admin.AdminEvent
+import org.keycloak.models.KeycloakSession
+import com.qupaya.util.urlEncode
+import java.io.IOException
+import java.net.URI
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+
+
+class NewsletterRegistrationEventListenerProvider(private val session: KeycloakSession) : EventListenerProvider {
+    private val brevoFormLink = System.getenv("BREVO_FORM_LINK")
+
+    init {
+        if (brevoFormLink.isNullOrEmpty()) {
+            LOG.warn("Environment variable BREVO_FORM_LINK is not set!")
+        }
+    }
+
+    override fun onEvent(event: Event?) {
+        if (event?.type != EventType.REGISTER) {
+            return
+        }
+
+        val user = session.users().getUserById(session.context.realm, event.userId)
+        if (user == null) {
+            LOG.warn("Unable to find user with ID ${event.id}")
+            return
+        }
+
+        val wantsNewsletter = user.attributes["newsletter"]
+        if (wantsNewsletter == null || !wantsNewsletter.contains("on")) {
+            return
+        }
+
+        val formData = mapOf(
+            "EMAIL" to user.email,
+            "OPT_IN" to "1",
+            "email_address_check" to "",
+            "locale" to "de",
+            "html_type" to "simple",
+        )
+
+        /*
+         * Yes, we could use the Brevo API to create a (DOI) contact, but
+         * using the form data POST request ensures that handling like
+         * success pages can be controlled inside Brevo.
+         */
+        val http = HttpClient.newHttpClient()
+        val request = HttpRequest.newBuilder()
+            .uri(URI.create(brevoFormLink))
+            .POST(HttpRequest.BodyPublishers.ofString(getFormDataAsString(formData)))
```

Comment:
don't we need error handling here? Should we maybe add a retry?

Reply:
done :)